omnipaxos/src/ballot_leader_election.rs

```diff
@@ -179,116 +199,86 @@ impl BallotLeaderElection {
         }
     }
 
-    pub(crate) fn hb_timeout(&mut self, seq_paxos_state: &(Role, Phase)) -> Option<Ballot> {
-        let mut tell_paxos = false;
-        let mut prints = vec![None, None, None];
+    pub(crate) fn hb_timeout(&mut self, seq_paxos_state: &(Role, Phase)) -> BleState {
+        self.check_leader();
+        self.update_happiness(seq_paxos_state);
+        self.update_ballot();
+        self.heartbeat_replies.clear();
+        self.new_hb_round();
+        self.state
+    }
 
-        let max_reply = self
+    /// Checks the heartbeat replies of the round and sets the new leader if appropriate
+    fn check_leader(&mut self) {
+        self.leader = self
             .heartbeat_replies
             .iter()
-            .max_by(|r1, r2| r1.ballot.cmp(&r2.ballot));
-        match max_reply {
-            Some(reply) if reply.ballot > self.leader.unwrap_or_default() => {
-                self.leader = Some(reply.ballot);
-                self.happy = reply.happy;
-                self.have_seen_a_quorum_of_followers = false;
-                prints[0] = Some(format!(
-                    "BLE {}: found a new leader {:?}",
-                    self.pid, self.leader
-                ));
-            }
-            _ => (),
+            .fold(self.leader, |a, b| a.max(b.ballot));
+        if self.leader != self.current_ballot {
+            self.state = BleState::Follower;
         }
+    }
 
-        // Update happiness
-        self.happy = if self.current_ballot == self.leader.unwrap_or_default() {
-            println!("BLE {}: checking happy for leader", self.pid);
-            let followers = self
-                .heartbeat_replies
-                .iter()
-                .filter(|hb_reply| hb_reply.leader == self.current_ballot)
-                .count()
-                + 1;
-            let happy = match seq_paxos_state {
-                (Role::Leader, Phase::Accept) => self.quorum.is_accept_quorum(followers),
-                _ => self.quorum.is_prepare_quorum(followers),
-            };
-            if happy && !self.have_seen_a_quorum_of_followers {
-                self.have_seen_a_quorum_of_followers = true;
-                tell_paxos = true;
+    fn update_happiness(&mut self, seq_paxos_state: &(Role, Phase)) {
+        self.happy = match self.state {
+            BleState::Follower => {
+                let leaders_reply = self
+                    .heartbeat_replies
+                    .iter()
+                    .find(|r| r.ballot.pid == self.leader.pid);
+                match leaders_reply {
+                    Some(reply) if reply.ballot >= self.leader => reply.happy,
+                    // Must have learned of leader ballot from HeartbeatRequest and HeartbeatReply is stale
+                    Some(_) => true,
```

Comment:
I understand the case you describe in the comment. But what if the leader crashes and comes back with a smaller (the default) ballot? Do we perhaps need a check for if the `reply.ballot` is less than `self.leader` and equals the default, then become unhappy?

Reply:
On a second thought, maybe the cleanest and most similar to the pseudo code is to use another variable `leaders_latest_reply` which is the max of `self.leader` and `leader_reply`. And then set happiness based on `leaders_latest_reply`

Reply:
Not sure I understand the second solution, but yeah it needs to be addressed. Another solution is to make ballots persistent.